### PR TITLE
feat(#8): Basic tokenizer

### PR DESCRIPTION
**Pull request by elenavoronin**

### CODE REVIEW

File: kotishell/src/tokenize/tokenize_a.c [hunk 1]
```diff
@@ -0,0 +1,90 @@
...
+/*   Created: 2023/06/29 16:20:23 by elenavoroni   #+#    #+#                 */
+/*   Updated: 2023/07/13 18:54:20 by dkolodze      ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "tokenize.h"
+
+static void	s_tk_whitespace(t_tk_so_far *so_far, char *s, t_tk_result *result)
+{
+	if (so_far->token.type != TK_WHITESPACE && so_far->token.length == 0)
```
Comment (Oktosha) [inline kotishell/src/tokenize/tokenize_a.c:17]: There can be 2 of same token in a row for pipes in our shell; maybe replace `so_far->token.length == 0` with `so_far->token.type == TK_EMPTY`

File: kotishell/src/tokenize/tokenize_a.c [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+	{
+		so_far->token.length = 1;
+		so_far->token.type = TK_WHITESPACE;
+		so_far->token.data = s;
+	}
+	else
+		so_far->token.length += 1;
+	if (tk_next_state(so_far->state, s + 1) != so_far->state)
```
Comment (Oktosha) [inline kotishell/src/tokenize/tokenize_a.c:25]: later more elaborate check needed

Comment (Oktosha) [reply kotishell/src/tokenize/tokenize_a.c:25]: probably will need to split tk_next_state into per-state deciding functions

File: kotishell/src/tokenize/tokenize_a.c [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+		else if (so_far.state == TK_ST_WORD)
+			s_tk_word(&so_far, s, &result);
+		else if (so_far.state == TK_ST_WHITESPACE)
+			s_tk_whitespace(&so_far, s, &result);
+		else if (so_far.state == TK_ST_END)
+			s_tk_end(&so_far, s, &result);
+		s++;
+		i--;
+	}
+	li_print_list(result.tokens);
```
Comment (Oktosha) [inline kotishell/src/tokenize/tokenize_a.c:88]: TODO: remove

File: kotishell/src/tokenize/tokenize_a.c [hunk 4]
```diff
@@ -0,0 +1,90 @@
...
+	t_tk_so_far		so_far;
+	int				i;
+
+	result.tokens = NULL;
+	s_tk_init_so_far(&so_far);
+	i = mini_strlen(s) + 1;
+	while (i >= 0)
+	{
+		if (so_far.state == TK_ST_START)
+			s_tk_start(&so_far, s, &result);
```
Comment (Oktosha) [inline kotishell/src/tokenize/tokenize_a.c:78]: may be move start and end out of the loop

File: kotishell/src/tokenize/tokenize_b.c [hunk 5]
```diff
@@ -0,0 +1,83 @@
...
+/*   Created: 2023/06/30 16:23:10 by evoronin      #+#    #+#                 */
+/*   Updated: 2023/07/12 16:26:35 by evoronin      ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "tokenize.h"
+
+t_tk_symbol_type	s_tk_get_symbol_type(char c)
+{
+	if (c == ' ' || c == '\t' || c == '\n' || c == '\v' || c == '\f')
```
Comment (Oktosha) [inline kotishell/src/tokenize/tokenize_b.c:17]: '\r' ?

File: kotishell/src/tokenize/tokenize_b.c [hunk 6]
```diff
@@ -0,0 +1,83 @@
...
+	{
+		so_far->token.length = 1;
+		so_far->token.type = TK_WORD;
+		so_far->token.data = s;
+	}
+	else
+		so_far->token.length += 1;
+	if (tk_next_state(so_far->state, s + 1) != so_far->state)
+	{
+		tk_token_copy(so_far, result);
```
Comment (Oktosha) [inline kotishell/src/tokenize/tokenize_b.c:46]: TODO: failed malloc

File: kotishell/src/tokenize/tokenize_a.c [hunk 7]
```diff
@@ -0,0 +1,90 @@
...
+	{
+		tk_token_copy(so_far, result);
+		so_far->token.length = 0;
+		so_far->state = tk_next_state(so_far->state, s + 1);
+	}
+}
+
+static void	s_tk_end(t_tk_so_far *so_far, char *s, t_tk_result *result)
+{
+	if (so_far->token.type != TK_EOL)
```
Comment (Oktosha) [inline kotishell/src/tokenize/tokenize_a.c:35]: add free on TK_ERROR_MALLOC / other problems

File: kotishell/src/tokenize/tokenize_b.c [hunk 8]
```diff
@@ -0,0 +1,83 @@
...
+{
+	so_far->status = TK_SUCCESS;
+	so_far->head = NULL;
+	so_far->token.type = TK_EMPTY;
+	so_far->token.length = 0;
+	so_far->state = TK_ST_START;
+	so_far->token.data = NULL;
+}
+
+void	s_tk_word(t_tk_so_far *so_far, char *s, t_tk_result *result)
```
Comment (Oktosha) [inline kotishell/src/tokenize/tokenize_b.c:34]: guard agains failed status

**Review by Oktosha — CHANGES_REQUESTED** [review]

File: kotishell/src/lists/lists_b.c [hunk 9]
```diff
@@ -0,0 +1,39 @@
...
+/*                                                   +#+                      */
+/*   Created: 2023/06/30 15:31:04 by evoronin      #+#    #+#                 */
+/*   Updated: 2023/07/12 11:54:09 by evoronin      ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "lists.h"
+#include "tokenize.h"
+
+void	li_list_free(t_li_node *list)
```
Comment (Oktosha) [inline kotishell/src/lists/lists_b.c:16]: need func parameter to properly free data

**Review by Oktosha — CHANGES_REQUESTED** [review]

File: kotishell/src/lists/lists_a.c [hunk 10]
```diff
@@ -0,0 +1,58 @@
...
+		list = li_new_list(data);
+	else
+		li_add_back(list, li_new_list(data));
+	return (list);
+}
+
+/**
+ * TODO: add proper assert function for our errors instead of exit(EXIT_FAILURE)
+*/
+t_li_node	*li_add_back(t_li_node *list, t_li_node *node)
```
[the remark this answers is not in the document]
Comment (elenavoronin) [reply kotishell/src/lists/lists_a.c:50]: you're right, changed!

File: kotishell/src/lists/lists_a.c [hunk 11]
```diff
@@ -0,0 +1,58 @@
...
+{
+	t_li_node	*node;
+
+	node = mini_malloc(sizeof(t_li_node));
+	node->data = data;
+	node->next = NULL;
+	return (node);
+}
+
+t_li_node	*li_new_stack(t_li_node *list, void *data)
```
[the remark this answers is not in the document]
Comment (elenavoronin) [reply kotishell/src/lists/lists_a.c:38]: I agree, especially on the return side, don't need to return anything

File: kotishell/src/minilibft/mini_strjoin.c [hunk 12]
```diff
@@ -0,0 +1,38 @@
...
+/*   By: evoronin <evoronin@student.codam.nl>         +#+                     */
+/*                                                   +#+                      */
+/*   Created: 2023/07/03 13:50:35 by evoronin      #+#    #+#                 */
+/*   Updated: 2023/07/03 14:30:03 by evoronin      ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "minilibft.h"
+
+char	*mini_strjoin(char *s1, char *s2, int len)
```
[the remark this answers is not in the document]
Comment (elenavoronin) [reply kotishell/src/minilibft/mini_strjoin.c:15]: yes!

File: kotishell/src/minilibft/mini_substr.c [hunk 13]
```diff
@@ -0,0 +1,38 @@
...
+#include "minilibft.h"
+
+char	*mini_substr(char *s1, unsigned int start, size_t len)
+{
+	char			*s2;
+	size_t			i;
+	size_t			size;
+
+	size = mini_strlen(s1);
+	if (s1 == NULL)
```
[the remark this answers is not in the document]
Comment (elenavoronin) [reply kotishell/src/minilibft/mini_substr.c:22]: removed! will rewrite for when I need it

File: kotishell/src/tokenize/tokenize_a.c (hunk 1 above)
Comment (elenavoronin) [reply kotishell/src/tokenize/tokenize_a.c:17]: agreed!

File: kotishell/src/tokenize/tokenize_b.c (hunk 5 above)
Comment (elenavoronin) [reply kotishell/src/tokenize/tokenize_b.c:17]: oopsie

File: kotishell/src/lists/lists_b.c (hunk 9 above)
Comment (elenavoronin) [reply kotishell/src/lists/lists_b.c:16]: moved to the tokenize module and made it token specific

File: kotishell/src/lists/lists_a.c [hunk 14]
```diff
@@ -0,0 +1,58 @@
...
+		list = list->next;
+	}
+	return (i);
+}
+
+t_li_node	*li_new_list(void *data)
+{
+	t_li_node	*node;
+
+	node = mini_malloc(sizeof(t_li_node));
```
[the remark this answers is not in the document]
Comment (elenavoronin) [reply kotishell/src/lists/lists_a.c:32]: please see trello board - we have to decide on this.

[the remark this answers is not in the document]
Comment (elenavoronin) [reply kotishell/src/lists/lists_a.c:32]: so far I'm going to return NULL and handle it in the other modules

File: kotishell/src/tokenize/tokenize_a.c (hunk 4 above)
Comment (elenavoronin) [reply kotishell/src/tokenize/tokenize_a.c:78]: done

File: kotishell/src/tokenize/tokenize_b.c (hunk 8 above)
Comment (elenavoronin) [reply kotishell/src/tokenize/tokenize_b.c:34]: done

File: kotishell/src/tokenize/tokenize_b.c (hunk 6 above)
Comment (elenavoronin) [reply kotishell/src/tokenize/tokenize_b.c:46]: done

File: kotishell/src/tokenize/tokenize_a.c (hunk 7 above)
Comment (elenavoronin) [reply kotishell/src/tokenize/tokenize_a.c:35]: done

File: kotishell/include/tokenize.h [hunk 15]
```diff
@@ -82,9 +82,12 @@ t_tk_result			tk_tokenize(char *s);
...
-t_tk_result			tk_result(char *s);
-void				tk_token_copy( t_tk_so_far *so_far, t_tk_result *result);
-void				tk_token_result(t_tk_result *result, t_tk_so_far *so_far);
-t_tk_state			tk_next_state(t_tk_state state, char *s);
+t_tk_result			s_tk_result(char *s);
+void				s_tk_token_copy( t_tk_so_far *so_far, t_tk_result *result);
+void				s_tk_token_result(t_tk_result *result, t_tk_so_far *so_far);
+t_tk_state			s_tk_next_state(t_tk_state state, char *s);
+void				s_tk_print_list(t_li_node *list);
+void				s_li_token_free(t_li_node *list);
```
Comment (Oktosha) [inline kotishell/include/tokenize.h:90]: `s_li_token_free` -> `l_tk_free_token_list`

File: kotishell/src/tokenize/tokenize_c.c [hunk 16]
```diff
@@ -25,3 +25,38 @@ t_tk_state	tk_next_state(t_tk_state state, char *s)
 		state = TK_ST_WORD;
 	return (state);
 }
+
+void	s_tk_print_list(t_li_node *list)
```
Comment (Oktosha) [inline kotishell/src/tokenize/tokenize_c.c:29]: Could make sense to move to the test directory

File: kotishell/src/minilibft/mini_putstr_fd.c [hunk 17]
```diff
@@ -0,0 +1,25 @@
...
+/*   By: elenavoronin <elnvoronin@gmail.com>          +#+                     */
+/*                                                   +#+                      */
+/*   Created: 2023/06/30 15:21:11 by evoronin      #+#    #+#                 */
+/*   Updated: 2023/07/04 17:04:10 by elenavoroni   ########   odam.nl         */
+/*                                                                            */
+/* ************************************************************************** */
+
+#include "minilibft.h"
+
+void	mini_putstr_fd(char *s, int fd)
```
[the remark this answers is not in the document]
Comment (Oktosha) [reply kotishell/src/minilibft/mini_putstr_fd.c]: For write:
+ we should not only check if it returned zero but also check if it wrote less than expected; we will think about it in the lifecycle

File: kotishell/src/tokenize/tokenize_a.c [hunk 18]
```diff
@@ -6,39 +6,42 @@
...
+	if (so_far->token.type == TK_EMPTY)
 	{
 		so_far->token.type = TK_EOL;
 		so_far->token.length = 0;
 		so_far->token.data = s;
 	}
-	tk_token_copy(so_far, result);
+	s_tk_token_copy(so_far, result);
+	if (so_far->status == TK_ERR_MALLOC)
+		s_tk_error(so_far, result);
```
Comment (Oktosha) [inline kotishell/src/tokenize/tokenize_a.c:44]: Let's call s_tk_error once only in tk_end; and add guard (check if the status is not TK_SUCCESS) that will prevent all the functions from doing things

File: kotishell/include/lists.h [hunk 19]
```diff
@@ -0,0 +1,36 @@
...
+{
+	void				*data;
+	struct s_li_node	*next;
+}	t_li_node;
+
+int			li_list_size(t_li_node *list);
+t_li_node	*li_new_list(void *data);
+void		li_new_stack(t_li_node **list, void *data);
+void		li_add_back(t_li_node **list, t_li_node *node);
+void		li_list_free(t_li_node *list);
```
Comment (Oktosha) [inline kotishell/include/lists.h:34]: Either add a way to provide a function to free the data or delete it: we can't use it the way it is, it doesn't free data and causes leaks

File: kotishell/src/lists/lists_a.c [hunk 20]
```diff
@@ -0,0 +1,61 @@
...
+/**
+ * TODO: add proper assert function for our errors instead of exit(EXIT_FAILURE)
+*/
+void	li_add_back(t_li_node **list, t_li_node *node)
+{
+	t_li_node	*new_list;
+
+	new_list = *list;
+	if (!new_list)
+		exit(EXIT_FAILURE);
```
Comment (Oktosha) [inline kotishell/src/lists/lists_a.c:57]: !new_list means we called the function in the wrong way; this error is our fault and should newer happen; We should write "BUG" with red color, and exit in this situation.

File: kotishell/src/lists/lists_a.c [hunk 21]
```diff
@@ -0,0 +1,61 @@
...
+void	li_new_stack(t_li_node **list, void *data)
+{
+	if (!(*list))
+		*list = li_new_list(data);
+	else
+		li_add_back(list, li_new_list(data));
+}
+
+/**
+ * TODO: add proper assert function for our errors instead of exit(EXIT_FAILURE)
```
Comment (Oktosha) [inline kotishell/src/lists/lists_a.c:49]: delete this todo; don't write todos next time; use review comments

File: kotishell/src/lists/lists_a.c [hunk 22]
```diff
@@ -0,0 +1,61 @@
...
+
+	node = mini_malloc(sizeof(t_li_node));
+	if (!node)
+		return (NULL);
+	node->data = data;
+	node->next = NULL;
+	return (node);
+}
+
+void	li_new_stack(t_li_node **list, void *data)
```
Comment (Oktosha) [inline kotishell/src/lists/lists_a.c:40]: Make it return status code: 0 if allocation worked, -1 otherwise;

File: kotishell/include/tokenize.h [hunk 23]
```diff
@@ -26,13 +31,63 @@ typedef enum e_token
 	TK_LESS,
 	TK_DGREAT,
 	TK_GREAT,
-}	t_token;
+	TK_EMPTY,
+	TK_WHITESPACE,
+	TK_EOL,
+}	t_tk_token_type;
+
+typedef enum e_tk_state
```
Comment (Oktosha) [inline kotishell/include/tokenize.h:39]: Add state TK_ST_ERROR

File: kotishell/src/tokenize/tokenize_a.c [hunk 24]
```diff
@@ -0,0 +1,91 @@
...
+	if (s_tk_next_state(so_far->state, s + 1) != so_far->state)
+	{
+		s_tk_token_copy(so_far, result);
+		so_far->token.type = TK_EMPTY;
+		so_far->token.length = 0;
+		so_far->state = s_tk_next_state(so_far->state, s + 1);
+	}
+}
+
+static void	s_tk_end(t_tk_so_far *so_far, char *s, t_tk_result *result)
```
Comment (Oktosha) [inline kotishell/src/tokenize/tokenize_a.c:34]:
```C
static void	s_tk_end(t_tk_so_far *so_far, char *s, t_tk_result *result)
{
	if (so_far->status == TK_SUCCESS)
	{
		so_far->token.type = TK_EOL;
		so_far->token.length = 0;
		so_far->token.data = s;
		s_tk_token_copy(so_far, result);
	}
	if (so_far->status != TK_SUCCESS)
		s_tk_error(so_far, result);
}
```

File: kotishell/src/tokenize/tokenize_a.c [hunk 25]
```diff
@@ -0,0 +1,91 @@
...
+	if (symbol == TK_SY_WHITESPACE)
+	{
+		so_far->state = TK_ST_WHITESPACE;
+		s_tk_whitespace(so_far, s, result);
+	}
+}
+
+t_tk_result	tk_tokenize(char *s)
+{
+	t_tk_result		result;
```
Comment (Oktosha) [inline kotishell/src/tokenize/tokenize_a.c:71]: You don't need to keep result updated; you can copy all the info from so_far in the end

**Review by Oktosha — CHANGES_REQUESTED** [review]

File: kotishell/src/tokenize/tokenize_c.c [hunk 26]
```diff
@@ -25,3 +25,38 @@ t_tk_state	tk_next_state(t_tk_state state, char *s)
...
+		temp = list;
+		free(temp->data);
+		list = list->next;
+		free(temp);
+	}
+}
+
+void	s_tk_error(t_tk_so_far *so_far, t_tk_result *result)
+{
+	s_li_token_free(so_far->head);
```
[the remark this answers is not in the document]
Comment (elenavoronin) [reply kotishell/src/tokenize/tokenize_c.c:59]: fixed

File: kotishell/include/tokenize.h (hunk 15 above)
Comment (elenavoronin) [reply kotishell/include/tokenize.h:90]: done

File: kotishell/src/tokenize/tokenize_c.c [hunk 27]
```diff
@@ -25,3 +25,38 @@ t_tk_state	tk_next_state(t_tk_state state, char *s)
...
+	while (list)
+	{
+		temp = list;
+		free(temp->data);
+		list = list->next;
+		free(temp);
+	}
+}
+
+void	s_tk_error(t_tk_so_far *so_far, t_tk_result *result)
```
[the remark this answers is not in the document]
Comment (elenavoronin) [reply kotishell/src/tokenize/tokenize_c.c:57]: changed to make it more clear

File: kotishell/src/tokenize/tokenize_c.c (hunk 16 above)
Comment (elenavoronin) [reply kotishell/src/tokenize/tokenize_c.c:29]: done

File: kotishell/src/lists/lists_a.c (hunk 21 above)
Comment (elenavoronin) [reply kotishell/src/lists/lists_a.c:49]: yep

File: kotishell/include/lists.h (hunk 19 above)
Comment (elenavoronin) [reply kotishell/include/lists.h:34]: deleted for now, we'll think about it when we need to

File: kotishell/src/tokenize/tokenize_a.c (hunk 24 above)
Comment (elenavoronin) [reply kotishell/src/tokenize/tokenize_a.c:34]: fixed

File: kotishell/src/tokenize/tokenize_a.c (hunk 25 above)
Comment (elenavoronin) [reply kotishell/src/tokenize/tokenize_a.c:71]: yep, updated

File: kotishell/src/lists/lists_a.c (hunk 20 above)
Comment (elenavoronin) [reply kotishell/src/lists/lists_a.c:57]: fixed

File: kotishell/include/tokenize.h (hunk 23 above)
Comment (elenavoronin) [reply kotishell/include/tokenize.h:39]: done

File: kotishell/src/lists/lists_a.c (hunk 22 above)
Comment (elenavoronin) [reply kotishell/src/lists/lists_a.c:40]: fixed in a different way, let's talk about it

**Review by Oktosha — APPROVED** [review]

**Pull request merged**